### ISSUE TITLE
simplify configuration if rust source was installed via rustup

### ISSUE THREAD

**Issue by colin-kiegel** [phildawes/racer#627]
# TL;DR

* add a fallback when `RUST_SRC_PATH` is undefined and `rustup` is installed

# background

rustup now comes with a way to install rust source code inside the root directory of each toolchain (https://github.com/rust-lang-nursery/rustup.rs/pull/659 and [#626](https://github.com/phildawes/racer/pull/626/files)). `rustup` can be queried which toolchain is active and which binaries are used, which points to the toolchain directory. And `rustc` itself can be queried for the current toolchain root directory.

```bash
$ rustup which rustc
/home/colin/.multirust/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc
$ rustc --print sysroot
/home/colin/.multirust/toolchains/stable-x86_64-unknown-linux-gnu
```

If the rust source code was installed via `rustup component add rust-src`, then it will reside in 
``` /home/colin/.multirust/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/ ```

This can be rewritten in terms of the above query commands as either:
* `$(rustup which rustc | xargs dirname)/../lib/rustlib/src/rust/src/`
* `$(rustc --print sysroot)/lib/rustlib/src/rust/src/`

A direct query would be nicer of course, but this is all we have now.

# suggestion

If `RUST_SRC_PATH` is undefined and `rustup` is installed,
then check if the source code is installed in the currently active toolchain (educated guess)
* If source code is installed, then use it
* else, give a hint like "please run `rustup component add rust-src` to install the rust source code"

# discussion

* more convenient first time setup
* more convenient and powerful on subsequent toolchain changes, since it will automagically use the source code of the currently active toolchain
* manual approach still possible

**Comment by colin-kiegel** [comment]
does this make sense to you?

**Comment by jwilm** [comment]
We've actually got something like this in Racer already. The [`check_rust_sysroot()`] function appears to be exactly what you're recommending. There's not a published version of it yet, but [that should be fixed soon].

Leaving this issue open as a reminder to fix docs

[`check_rust_sysroot()`]: https://github.com/phildawes/racer/blob/c680530f9e9dcd8de5ef0d45954a6112a01a6fe5/src/bin/main.rs#L169-L183
[that should be fixed soon]: https://github.com/phildawes/racer/issues/619

**Comment by colin-kiegel** [comment]
Ah cool. Btw, I also sent a pr to update the docs #626

**Comment by jwilm** [comment]
Oh, great! Guess this can be closed then. Thanks for the write up :smile:

**Comment by colin-kiegel** [comment]
Well, if `check_sys_root()` lands, the docs should be updated again, because setting the environment variable wouldn't be necessary with the dust-up approach anymore, right?

Apart from that I am happy that this in the making. :-)
